src/Agent.Listener/Configuration/WindowsSecurityManager.cs

```diff
@@ -0,0 +1,477 @@
+        private const int UnicodeCharSize = 2;
+
+        private static LSA_UNICODE_STRING[] LogonAsServiceRights
+        {
+            get
+            {
+                return new[]
+                           {
+                               new LSA_UNICODE_STRING()
+                                   {
+                                       Buffer = Marshal.StringToHGlobalUni(s_logonAsServiceName),
+                                       Length = (UInt16)(s_logonAsServiceName.Length * UnicodeCharSize),
+                                       MaximumLength = (UInt16) ((s_logonAsServiceName.Length + 1) * UnicodeCharSize)
+                                   }
+                           };
+            }
+        }
+
+        public struct ReturnCode
+        {
+            public const int S_OK = 0;
+            public const int ERROR_ACCESS_DENIED = 5;
+            public const int ERROR_INVALID_PARAMETER = 87;
+            public const int ERROR_MEMBER_NOT_IN_ALIAS = 1377; // member not in a group            
+            public const int ERROR_MEMBER_IN_ALIAS = 1378; // member already exists
+            public const int ERROR_ALIAS_EXISTS = 1379;  // group already exists
+            public const int ERROR_NO_SUCH_ALIAS = 1376;
+            public const int ERROR_NO_SUCH_MEMBER = 1387;
+            public const int ERROR_INVALID_MEMBER = 1388;
+            public const int NERR_GroupNotFound = 2220;
+            public const int NERR_GroupExists = 2223;
+            public const int NERR_UserInGroup = 2236;
+        }
+
+        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+        public struct LocalGroupInfo
+        {
+            [MarshalAs(UnmanagedType.LPWStr)]
+            public string Name;
+            [MarshalAs(UnmanagedType.LPWStr)]
+            public string Comment;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        public struct LSA_UNICODE_STRING
+        {
+            public UInt16 Length;
+            public UInt16 MaximumLength;
+
+            // We need to use an IntPtr because if we wrap the Buffer with a SafeHandle-derived class, we get a failure during LsaAddAccountRights
+            public IntPtr Buffer;
+        }
+
+        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+        public struct LocalGroupMemberInfo
+        {
+            [MarshalAs(UnmanagedType.LPWStr)]
+            public string FullName;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        public struct LSA_OBJECT_ATTRIBUTES
+        {
+            public UInt32 Length;
+            public IntPtr RootDirectory;
+            public LSA_UNICODE_STRING ObjectName;
+            public UInt32 Attributes;
+            public IntPtr SecurityDescriptor;
+            public IntPtr SecurityQualityOfService;
+        }
+
+        [DllImport("Netapi32.dll")]
+        private extern static int NetLocalGroupGetInfo(string servername,
```

Comment:
I spent a few minutes looking into how to deal with local group membership in the full .Net Framework. It looks like there's classes in the namespace `System.DirectoryServices.AccountManagement`. I don't know the history why these classes weren't used in the old agent. The Win32 APIs may have been more reliable or maybe there is already so much related Win32 APIs being called that it makes sense anyway.

Anyway I couldn't find any classes under the namespace `System.DirectoryServices.AccountManagement` @ https://dotnet.github.io/api/ . Not sure if there's some other classes available. I couldn't find anything but only searched briefly.

Doesn't really matter either way. Just thought I'd share what I found.